server/src/main/scala/com/cloudera/livy/server/LivyServer.scala

```diff
@@ -135,6 +161,40 @@ class LivyServer extends Logging {
     sys.props("livy.server.serverUrl") = _serverUrl.get
   }
 
+  def runKinit(livyConf: LivyConf): Unit = {
+    val keytab = livyConf.get(LAUNCH_KERBEROS_KEYTAB)
+    val principal = livyConf.get(LAUNCH_KERBEROS_PRINCIPAL)
+    require(principal != null,
+      s"Kerberos requires ${LAUNCH_KERBEROS_KEYTAB.key} to be provided.")
+    require(keytab != null,
+      s"Kerberos requires ${LAUNCH_KERBEROS_PRINCIPAL.key} to be provided.")
+    val cache_name = livyConf.get(KINIT_CACHE_NAME)
+    val commands = Seq("kinit", "-kt", keytab, "-c", cache_name, principal)
+    val proc = new ProcessBuilder(commands: _*).inheritIO().start()
+    proc.waitFor() match {
+      case 0 =>
+        debug("Ran kinit command successfully.")
+        kinitFailCount = 0
+        UserGroupInformation.getCurrentUser.reloginFromTicketCache()
+      case _ =>
+        warn("Fail to run kinit command.")
+        kinitFailCount += 1
+        val kinitFailThreshold = livyConf.getInt(KINIT_FAIL_THRESHOLD)
+        if (kinitFailCount >= kinitFailThreshold) {
+          error(s"Exit LivyServer after ${kinitFailThreshold} times failures running kinit.")
+          if (server.server.getState == "STARTED") {
+            stop()
+          } else {
+            sys.exit(1)
+          }
+        } else {
+          // retry after 5 seconds
+          Thread.sleep(5*1000)
```

Comment:
This is a little weird. I'd just schedule a one-off task to run `runKinit` using the existing executor. Which required addressing my previous feedback of making `executor` a class field.